mobile/android/app/src/main/kotlin/com/example/mobile/BackupWorker.kt

```diff
@@ -0,0 +1,306 @@
+import androidx.work.OneTimeWorkRequest
+import androidx.work.WorkManager
+import com.google.common.util.concurrent.ListenableFuture
+import io.flutter.embedding.engine.FlutterEngine
+import io.flutter.embedding.engine.dart.DartExecutor
+import io.flutter.embedding.engine.loader.FlutterLoader
+import io.flutter.plugin.common.MethodCall
+import io.flutter.plugin.common.MethodChannel
+import io.flutter.view.FlutterCallbackInformation
+import java.util.concurrent.TimeUnit
+
+/**
+ * Worker executed by Android WorkManager to perform backup in background
+ *
+ * Starts the Dart runtime/engine and calls `_nativeEntry` function in
+ * `background.service.dart` to run the actual backup logic.
+ * Called by Android WorkManager when all constraints for the work are met,
+ * i.e. a new photo/video is created on the device AND battery is not low.
+ * Optionally, unmetered network (wifi) and charging can be required.
+ * As this work is not triggered periodically, but on content change, the
+ * worker enqueues itself again with the same settings.
+ * In case the worker is stopped by the system (e.g. constraints like wifi
+ * are no longer met, or the system needs memory resources for more other
+ * more important work), the worker is replaced without the constraint on
+ * changed contents to run again as soon as deemed possible by the system.
+ */
+class BackupWorker(ctx: Context, params: WorkerParameters) : ListenableWorker(ctx, params), MethodChannel.MethodCallHandler {
+
+    private val resolvableFuture = ResolvableFuture.create<Result>()
+    private var engine: FlutterEngine? = null
+    private lateinit var backgroundChannel: MethodChannel
+    private val notificationManager = ctx.getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager
+    private val isIgnoringBatteryOptimizations = isIgnoringBatteryOptimizations(applicationContext)
+
+    override fun startWork(): ListenableFuture<ListenableWorker.Result> {
+
+        val ctx = applicationContext
+        // enqueue itself once again to continue to listen on added photos/videos
+        enqueueMoreWork(ctx,
+                        requireUnmeteredNetwork = inputData.getBoolean(DATA_KEY_UNMETERED, true),
+                        requireCharging = inputData.getBoolean(DATA_KEY_CHARGING, false))
+
+        if (!flutterLoader.initialized()) {
+            flutterLoader.startInitialization(ctx)
+        }
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
+            // Create a Notification channel if necessary
+            createChannel()
+        }
+        if (isIgnoringBatteryOptimizations) {
+            // normal background services can only up to 10 minutes
+            // foreground services are allowed to run indefinitely
+            // requires battery optimizations to be disabled (either manually by the user
+            // or by the system learning that immich is important to the user)
+            setForegroundAsync(createForegroundInfo())
```

Comment:
If a new asset is added to an excluded album, `onPhotosChanged` never gets called and a blank notification is shown for a few seconds. Maybe you could add default content (something like "Checking for new assets...").

Reply:
That's a good idea!
On my virtual devices, Android shows the foreground notification only after ~10 seconds, so this should not happen usually. However, some older devices might show it sooner.
Note: `onPhotosChanged` is called regardless of which albums were changed